src/meshes/t8code_mesh.jl

```diff
@@ -0,0 +1,338 @@
+@inline ninterfaces(mesh::T8codeMesh) = mesh.ninterfaces
+@inline nmortars(mesh::T8codeMesh) = mesh.nmortars
+@inline nboundaries(mesh::T8codeMesh) = mesh.nboundaries
+
+function Base.show(io::IO, mesh::T8codeMesh)
+    print(io, "T8codeMesh{", ndims(mesh), ", ", real(mesh), "}")
+end
+
+function Base.show(io::IO, ::MIME"text/plain", mesh::T8codeMesh)
+    if get(io, :compact, false)
+        show(io, mesh)
+    else
+        setup = [
+            "#trees" => ntrees(mesh),
+            "current #cells" => ncells(mesh),
+            "polydeg" => length(mesh.nodes) - 1,
+        ]
+        summary_box(io,
+                    "T8codeMesh{" * string(ndims(mesh)) * ", " * string(real(mesh)) * "}",
+                    setup)
+    end
+end
+
+"""
+    T8codeMesh(trees_per_dimension; polydeg, mapping=identity,
+               RealT=Float64, initial_refinement_level=0, periodicity=true)
+
+Create a structured potentially curved 'T8codeMesh' of the specified size.
+
+Non-periodic boundaries will be called ':x_neg', ':x_pos', ':y_neg', ':y_pos', ':z_neg', ':z_pos'.
+
+# Arguments
+- 'trees_per_dimension::NTupleE{NDIMS, Int}': the number of trees in each dimension.
+- 'polydeg::Integer': polynomial degree used to store the geometry of the mesh.
+                      The mapping will be approximated by an interpolation polynomial
+                      of the specified degree for each tree.
+- 'mapping': a function of 'NDIMS' variables to describe the mapping that transforms
+             the reference mesh ('[-1, 1]^n') to the physical domain.
+- 'RealT::Type': the type that should be used for coordinates.
+- 'initial_refinement_level::Integer': refine the mesh uniformly to this level before the simulation starts.
+- 'periodicity': either a 'Bool' deciding if all of the boundaries are periodic or an 'NTuple{NDIMS, Bool}'
+                 deciding for each dimension if the boundaries in this dimension are periodic.
+"""
+function T8codeMesh(trees_per_dimension; polydeg,
+                    mapping = coordinates2mapping((-1.0, -1.0), (1.0, 1.0)),
+                    RealT = Float64, initial_refinement_level = 0, periodicity = true)
+    NDIMS = length(trees_per_dimension)
+
+    @assert NDIMS == 2 # Only support for NDIMS = 2 yet.
+
+    # Convert periodicity to a Tuple of a Bool for every dimension
+    if all(periodicity)
+        # Also catches case where periodicity = true
+        periodicity = ntuple(_ -> true, NDIMS)
+    elseif !any(periodicity)
+        # Also catches case where periodicity = false
+        periodicity = ntuple(_ -> false, NDIMS)
+    else
+        # Default case if periodicity is an iterable
+        periodicity = Tuple(periodicity)
+    end
+
+    conn = T8code.Libt8.p4est_connectivity_new_brick(trees_per_dimension..., periodicity...)
+    do_partition = 0
+    cmesh = t8_cmesh_new_from_p4est(conn, mpi_comm(), do_partition)
+    T8code.Libt8.p4est_connectivity_destroy(conn)
```

Comment:
There were issues with using `p4est_connectivity_new_brick` etc. from P4est.jl, correct? Could be worth a comment here.

Reply:
Not sure what you mean. There were some issues on reading p4est connectivity structures in t8code which were resolved some time ago.